### PR TITLE
[AC 7642] P1: Users unable to access some people directory profiles

### PR DESCRIPTION
**Pull request by iankigen**
## [AC 7642](https://masschallenge.atlassian.net/browse/AC-7642)

**Changes introduced**
- Implement access control for people directory profiles

**Testing**
- Check out this branch 
- Login or masquerade as a user with the current user role of [Staff, Finalist, Alumni in residence, Mentor, Partner, Alum or No user role]
- Visit the people directory pages: `/people/<user_id>` of the users they can view and confirm that access control is as defined in the table below:
Screenshot 2020-04-15 at 17 19 27

### CODE REVIEW

File: web/impact/impact/graphql/query.py
Comment (jonkiparsky) [inline web/impact/impact/graphql/query.py]: Why are we making this into a list here?

If we leave it as a queryset, then we do not need to convert it back into a set on line 36

File: web/impact/impact/graphql/query.py
Comment (jonkiparsky) [inline web/impact/impact/graphql/query.py]: This query does not get current user roles, it gets all user roles that this user holds.

Presumably you could fix this by filtering `user.programrolegrant_set` on program_role.program.program_status

Probably we'd want to write a test about this, also, since it seems like an easy mistake to make. 
test should ensure that a user with a certain user role for an inactive program does not get to see things that only active grant holders should see.

File: web/impact/impact/tests/test_graphql.py
Comment (jonkiparsky) [inline web/impact/impact/tests/test_graphql.py]: we do not need to use - and prefer not to use - the `.create` syntax, since `ProgramRoleFactory(user_role=user_role)` means and does exactly the same thing.

**Review by jonkiparsky — CHANGES_REQUESTED** [review]
Some changes requested

**Review by freshprincekla — APPROVED** [review]
QA checks out. 
- Was able to login as the various users and verify that they have access to the correct profiles.

File: web/impact/impact/graphql/query.py
Comment (iankigen) [reply web/impact/impact/graphql/query.py]: I'm making `current_user_roles` a list so that I can perform a set intersection operation on `basic_user_roles`.
I'm converting `basic_user_roles` to set, not `current_user_roles`

**Comment by iankigen** [comment]
Working on feedback

**Comment by iankigen** [comment]
@jonkiparsky PR ready for a second review

File: web/impact/impact/tests/test_graphql.py
Comment (jonkiparsky) [inline web/impact/impact/tests/test_graphql.py]: Not a blocker for this ticket, but if we have multiple assertions in a single test (including assertions in a loop) this indicates that this is testing multiple things and therefore should be multiple tests. 

In future, please write individual tests for individual assertions. (and if you see tests like this, please fix them!)

File: web/impact/impact/tests/test_graphql.py
Comment (jonkiparsky) [inline web/impact/impact/tests/test_graphql.py:689]: It looks like this repeated definition indicates that we have a common pattern which should have been extracted. Hopefully this will be caught and cleaned up the next time we touch this test case, not going to block on this today.

File: web/impact/impact/permissions/graphql_permissions.py
Comment (jonkiparsky) [inline web/impact/impact/permissions/graphql_permissions.py]: This function is longer than our standards call for. 
In future, consider taking the following steps to get functions under 10 lines

- declare constants outside of functions (this is a benefit, since it allows them to be referred to in tests)
- extract complex queries to helper functions (allowing them to be re-used, and also granting them a helpful name)

File: web/impact/impact/permissions/graphql_permissions.py
Comment (jonkiparsky) [inline web/impact/impact/permissions/graphql_permissions.py]: In future, consider using a construction like

```
if any([role in basic_user_roles for role in current_logged_in_user_roles]):
   ....
```
instead of set intersections, to help readability. This will allow the reader to focus more directly on the logic.

**Review by jonkiparsky — COMMENTED** [review]
@iankigen @hfreund QA note: on cleandb, I signed in as user 76104-user@example.com, who has a Judge program role grant in an active program. This user was unable to access http://localhost:1234/people/

Therefore I was unable to verify that this user was able to see the correct users.

**Comment by iankigen** [comment]
@jonkiparsky this ticket is only focused on the people profile page, I believe since [http://localhost:1234/people/](http://localhost:1234/people/) is not the only entry point for the people profile pages.

**Comment by jonkiparsky** [comment]
@iankigen If you can provide better QA directions, that would be helpful. I would have assumed that access to the people directory meant access to  /people/ - is this not so?

**Comment by jonkiparsky** [comment]
@iankigen working as user 15-user@example.com (no user roles at all) I was able to access http://localhost:1234/directory/people/20863, which is a non-staff user. This seems inconsistent with the QA chart.

**Review by jonkiparsky — CHANGES_REQUESTED** [review]
QA fail: working as user 15-user@example.com (no user roles at all) I was able to access http://localhost:1234/directory/people/20863, which is a non-staff user. This is inconsistent with the QA chart, which says that users with no roles should only be able to access staff cards.

**Comment by iankigen** [comment]
Resolving conflicts

**Comment by hfreund** [comment]
@iankigen can you deploy this to a test machine so I can run through some QA as well?

**Review by jonkiparsky — COMMENTED** [review]
I have been unable to do a satisfactory QA pass. This work has been deployed to a test machine for Product to review, will approve when Hugh gives the okay

**Comment by hfreund** [comment]
@iankigen while doing QA on test2 I found the following problems:
Masquerading as https://test2.masschallenge.org/admin/simpleuser/user/49509/masquerade/
1. I removed the PRG for confirmed mentor so the user had only confirmed judge roles
2. Was unable to view https://test2.masschallenge.org/people/18223/ who is a 2017 Finalist for Boston (as a superuser I am able to view this user)
3. Added a confirmed mentor role grant to user 49509 and attempted again to view https://test2.masschallenge.org/people/18223/; was again unsuccessful
4. I was successfully able to view https://test2.masschallenge.org/people/78199/ who is an **alum.**
It appears users with the current program Global Alumni will appear in the People Directory filter for Global Alumni even if they do not have the Alum role grant. This seems like a separate issue but may have contributed to some of the challenges in QAing this ticket (the profiles listed on the original ticket all appear to have this data issue).

Masquerading as https://test2.masschallenge.org/admin/simpleuser/user/1277/masquerade/ which is a member profile type and has only partner admin role grants I was unable to view  https://test2.masschallenge.org/people/18223/ which is the expected outcome, although based on the failures detailed above I'm not confident if this block is due to an appropriate measure in the code or by some accident that may also be causing the access issue detailed in points 103 above (possibly this is because 18223 is a user in a program that is ended and therefore no one can see them, except super users). If that is the case the issue is a data one: user 18223 was not tagged as a global alum.

While still masquerading as 1277, I can navigate directly to https://test2.masschallenge.org/people/ (unclear if this should be possible since a member profile with a partner-admin user role does not have a defined visibility setting) however when I click on any profile I get an error.

@jonkiparsky I don't have strong feelings against merging this code since it does not appear to create undesired outcomes, so if you want to merge it please do so. However, this additional QA has revealed more problems and unsupported use-cases that should be addressed in follow on work:
1. Investigate why so many users with the current program **Global Alumni** do not have an **Alum** role grant
2. Investigate why the directory filter for Program uses **Current program** instead of the role grants a user has. This leads to unintended and inconsistent results and access
3. Investigate why a profile which cannot be accessed still appears in search results
4. Simplify access to directories, directory search results, and profile pages. Consider how different roles, user types (staff, finalist, expert, member), and program status does or should impact access

**Comment by iankigen** [comment]
@hfreund https://test2.masschallenge.org/admin/simpleuser/user/49509/change/ is not a staff user

**Review by jonkiparsky — APPROVED** [review]

**Comment by hfreund** [comment]
> @hfreund https://test2.masschallenge.org/admin/simpleuser/user/49509/change/ is not a staff user

@iankigen Correct. They should be able to see user 18223 but cannot (I am able to see them as a staff user). My suspicion is that 18223 should have been tagged as an Alum but was not for some reason (possibly due to oversight, possibly due to a broken process). This is one of the tickets that I will create to investigate.

**Pull request merged**
